nevergrad/optimization/recaster.py

```diff
@@ -206,3 +207,77 @@ class SequentialRecastOptimizer(RecastOptimizer):
     # pylint: disable=abstract-method
 
     no_parallelization = True
+
+
+class BatchRecastOptimizer(RecastOptimizer):
+    """Recast optimizer where points to evaluate are provided in batches
+    and stored by the optimizer to be asked and told on. The fake_callable
+    is only brought into action every 'batch size' number of asks and tells
+    instead of every ask and tell. This opens up the optimizer to
+    parallelism.
+
+    Note
+    ----
+    These implementations are not necessarily robust. You have to complete
+    a batch before you start a new one so parallelism is only possible
+    within batches i.e. if a batch size is 100 and you have done 100 asks,
+    you must do 100 tells before you ask again but you could do those 100
+    asks and tells in parallel.
+
+    """
+
+    # pylint: disable=abstract-method
+
+    def __init__(
+        self, parametrization: IntOrParameter, budget: tp.Optional[int] = None, num_workers: int = 1
+    ) -> None:
+        super().__init__(parametrization, budget, num_workers=num_workers)
+        self._current_batch: tp.List[p.Parameter] = []
+        self._batch_losses: tp.List[tp.Loss] = []
+        self._tell_counter = 0
+        self.batch_size = 0
+        self.indices: tp.Dict[p.Parameter, int] = {}
+
+    def _internal_ask_candidate(self) -> p.Parameter:
+        """Reads messages from the thread in which the underlying optimization function is running
+        New messages are sent as "ask".
+        """
+        # get a datapoint that is a random point in parameter space
+        if self._messaging_thread is None:
+            self._messaging_thread = MessagingThread(self.get_optimization_function())
+        # wait for a message
+        if not self._messaging_thread.is_alive():  # In case the algorithm stops before the budget is elapsed.
+            warnings.warn(
+                "Underlying optimizer has already converged, returning random points",
+                base.errors.FinishedUnderlyingOptimizerWarning,
+            )
+            self._check_error()
+            data = self._rng.normal(0, 1, self.dimension)
+            return self.parametrization.spawn_child().set_standardized_data(data)
+        # if no more points left in batch, wait for new batch from fake callable
+        if not self._current_batch:
+            points = self._messaging_thread.messages_ask.get()
+            self.batch_size = len(points)
+            self._current_batch = [self.parametrization.spawn_child(point) for point in points]
+            self._batch_losses = [None] * len(points)  # type: ignore
+            # map each point to an index in preparation to build loss array in tell
+            self.indices = {candidate: i for i, candidate in enumerate(self._current_batch)}
+        candidate = self._current_batch.pop()
```

Comment:
what happens if we pull too many candidates? An explicit error explaining that the batch size is too small could be helpful. Note that some algorithms work in steady mode, i.e if you just did a tell, you can always do a new ask. That won't be the case here so some users could get a bit lost and may need to be explained what happened ;)

Reply:
If you ask before the previous batch has been fully told on you end up in a deadlock potentially (if more than one worker asks) since you have multiple workers blocking to get from a queue that will only have one member. I have added a descriptive runtime error and a new function `can_ask()` that handles the logic of knowing whether an ask can be performed.

Reply:
`
if not self.can_ask(): raise RuntimeError( "You can't get a new batch until the old one has been fully told on. See docstring for more info." )
`
Instead of the batch being too small I have added an error stating that a premature ask has occurred.